modules/primo/primo.module

```diff
@@ -0,0 +1,201 @@
+function primo_ting_covers($entities) {
+  $ids = array_map(function(TingEntity $entity) {
+    return $entity->ding_entity_id;
+  }, $entities);
+
+  // Replace entity ids with a set of Primo ids which are known to have covers.
+  // TODO: Remove this once we have proper search integration with Primo.
+  $ids = array(
+    'ICE01_PRIMO_TEST000811009',
+    'ICE01_PRIMO_TEST001027834',
+    'ICE01_PRIMO_TEST000335899',
+  );
+
+  $document = primo_load_by_recordid($ids);
+
+  $thumbnail_urls = array_fill_keys($ids, NULL);
+  array_walk($thumbnail_urls, function(&$url, $id, DOMDocument $document) {
+    $urls = primo_extract_thumbnail_urls($document, $id);
+    // We can only handle a single thumbnail per id so return the first.
+    $url = array_pop($urls);
+  }, $document);
+
+  // Splice the results back into an array mapping original ids with thumbnail
+  // urls for Primo ids.
+  // TODO: Remove this once we have proper search integration with Primo.
+  $thumbnail_urls = array_slice($thumbnail_urls, 0, count($entities));
+  return array_combine(
+    array_keys($entities),
+    array_pad($thumbnail_urls, count($entities), reset($thumbnail_urls))
+  );
+}
+
+/**
+ * Load documents based on their record id.
+ *
+ * @param string[] $rids
+ *   An array of Primo record ids.
+ *
+ * @return \DOMDocument
+ *   A DOMDocument containing the resulting XML. This XML will contain elements
+ *   for each of the requested record ids.
+ */
+function primo_load_by_recordid(array $rids) {
+  $url = variable_get('primo_base_url') . '/PrimoWebServices/xservice/search/brief';
+  $params = array(
+    'institution' => variable_get('primo_institution_code'),
+    'query' => 'rid,contains,' . implode($rids, ' OR '),
+    'indx' => 1,
+    'bulkSize' => count($rids),
+  );
+
+  $result = drupal_http_request(url($url, array('query' => $params)));
+
+  $document = new DOMDocument();
+  $document->loadXML($result->data);
+
+  return $document;
+}
+
+/**
+ * Extract thumbnail urls for a record in an XML document
+ *
+ * @param \DOMDocument $document
+ *   The document containing record information.
+ * @param string $id
+ *   The record id for which to extract the thumbnail url.
+ * @return string[]
+ *   An array of thumbnail urls for the record.
+ */
+function primo_extract_thumbnail_urls(DOMDocument $document, $id) {
+  $xpath = new DOMXPath($document);
+  $xpath->registerNamespace(
+    's', 'http://www.exlibrisgroup.com/xsd/jaguar/search');
+  $xpath->registerNamespace(
+    'p', 'http://www.exlibrisgroup.com/xsd/primo/primo_nm_bib');
+  $thumbnail_elements = $xpath->query('//s:DOC[.//p:recordid[.="' . $id .  '"]]//s:thumbnail'
+  );
+
+  $thumbnail_urls = [];
+  foreach ($thumbnail_elements as $thumbnailElement) {
+    $thumbnail_urls[] = $thumbnailElement->nodeValue;
+  }
+
+  // Primo results are known to include duplicates. Remove these.
+  $thumbnail_urls = array_unique($thumbnail_urls);
+
+  $thumbnail_urls = array_filter($thumbnail_urls, function ($url) {
```

Comment:
Does the order of the urls have any significance? Mostly curious as I suspect we can't really affect what primo returns?

Reply:
I do not know. Primo usually returns 4 urls (2 unique) - one from `leitir.is` and one from Google Books.